Merge DID features into dev

matt-o-how opened the pull request.

matt-o-how committed:
start fixed rebase

matt-o-how committed:
fixed puzzle opcodes

matt-o-how committed:
changed (q to (q .
changed hardcoded puzzle

Yostra committed:
tx block

Yostra committed:
fix puzzles / did stuff

matt-o-how committed:
update innerpuz currryargs

Yostra committed:
farm full block

Yostra committed:
formatting

Yostra committed:
tests pass

matt-o-how committed:
fix superlint errors

mariano54 commented on src/wallet/wallet_state_manager.py:
```diff
@@ -93,19 +95,22 @@ class WalletStateManager:
         self.new_wallet = False
         self.config = config
         self.constants = constants
+        self.server = server
```
What is the purpose of this? This adds a circular dependency

mariano54 commented on src/util/streamable.py:
```diff
@@ -65,7 +65,13 @@ def dataclass_from_dict(klass, d):
+        # Type is tuple, can have multiple different types inside
+        i = 0
+        klass_properties = []
+        for item in d:
```
Can we add a test for the streamable changes? This is very sensitive code

mariano54 replied:
BTW we need to be careful with this, because we now have streamable in production, so we need to make sure nothing breaks.

matt-o-how replied:
The previous code assumed that Tuples only contained one type, this code checks each entry's type individually

matt-o-how committed:
black formatted did_wallet_puzzles

matt-o-how committed:
make creation records more like CC creation

matt-o-how committed:
added get_max_send_amount for DID

matt-o-how committed:
added catch for create DID wallet in RPC

hoffmang9 commented:
I'd like to try to get this in. Maybe rebase and see what issues are actually left after RC5?

Yostra committed:
cleanup, branch

Yostra committed:
did wallet, remove ((a replace with (a

matt-o-how committed:
fixes for clvm compilation tests

matt-o-how committed:
simplified max send because wallet is single coin

hoffmang9 reviewed (dismissed):
+1

matt-o-how committed:
Merge branch 'dev' into fixed_did_rebase

matt-o-how committed:
Merge branch 'dev' into fixed_did_rebase

matt-o-how committed:
lint fix

matt-o-how committed:
remove duplicate code

fixed potential bug where wallet_state_manager might not create DID 
wallet from SQL

matt-o-how committed:
reduce test length
include attest files in gitignore

matt-o-how committed:
re-introduce 5 blocks per test

matt-o-how committed:
remove potential attack highlighted by quexington

matt-o-how committed:
added some sleep calls for debugging purpose

matt-o-how committed:
attempt to fix tests

matt-o-how committed:
black fix for linter

matt-o-how committed:
add more value asserts for safety

matt-o-how committed:
allow extra time for wallet tests

justinengland commented:
can you please resubmit this PR to either main, or if it needs to be merged before 3/24/21, new-main, this will allow us to move forward with the new git branching strategy

matt-o-how committed:
used safe hexstr_to_bytes in a number of places in RPC API

matt-o-how committed:
stricter typing for DID

justinengland commented:
moved to main

matt-o-how committed:
Optimise DID_Core

matt-o-how committed:
change DID_CORE to SINGLETON_TOP_LAYER with new rules

matt-o-how committed:
black formatted singleton test
